src/sentry/tagstore/snuba/backend.py

```diff
@@ -16,69 +17,363 @@
+        start, end = self.get_time_range()
+        tag = 'tags[{}]'.format(key)
+        filters = {
+            'project_id': [project_id],
+            'environment': [environment_id],
+        }
+        if group_id is not None:
+            filters['issue'] = [group_id]
+        conditions = [[tag, '!=', '']]
+        aggregations = [['count', '', 'count']]
+
+        result = snuba.query(start, end, [], conditions, filters, aggregations)
+        if result == 0:
+            raise GroupTagKeyNotFound
+        else:
+            return ObjectWrapper({
+                'times_seen': result,
+                'key': key,
+                'group_id': group_id,
+            })
 
     def get_group_tag_keys(self, project_id, group_id, environment_id, limit=None):
-        pass
+        limit = limit or 1000
+        start, end = self.get_time_range()
+        filters = {
+            'project_id': [project_id],
+            'environment': [environment_id],
+        }
+        if group_id is not None:
+            filters['issue'] = [group_id]
+        aggregations = [['count', '', 'count']]
+
+        result = snuba.query(start, end, ['tags.key'], [], filters,
+                             aggregations, limit=limit, orderby='-count', arrayjoin='tags')
+
+        return [ObjectWrapper({
+            'times_seen': count,
+            'key': name,
+            'group_id': group_id,
+        }) for name, count in six.iteritems(result)]
 
     def get_group_tag_value(self, project_id, group_id, environment_id, key, value):
-        pass
+        from sentry.tagstore.exceptions import GroupTagValueNotFound
+        start, end = self.get_time_range()
+        tag = 'tags[{}]'.format(key)
+        filters = {
+            'project_id': [project_id],
+            'environment': [environment_id],
+        }
+        if group_id is not None:
+            filters['issue'] = [group_id]
+        conditions = [
+            [tag, '=', value]
+        ]
+        aggregations = [['count', '', 'count']]
+
+        result = snuba.query(start, end, [], conditions, filters, aggregations)
+
+        if result == 0:
+            raise GroupTagValueNotFound
+        else:
+            return ObjectWrapper({
+                'times_seen': result,
+                'key': key,
+                'value': value,
+                'group_id': group_id,
+            })
 
     def get_group_tag_values(self, project_id, group_id, environment_id, key):
-        pass
+        start, end = self.get_time_range(7)
+        tag = 'tags[{}]'.format(key)
+        filters = {
+            'project_id': [project_id],
+            'environment': [environment_id],
+        }
+        if group_id is not None:
+            filters['issue'] = [group_id]
+        conditions = [[tag, '!=', '']]
+        aggregations = [
+            ['count', '', 'count'],
+            ['min', SEEN_COLUMN, 'first_seen'],
+            ['max', SEEN_COLUMN, 'last_seen'],
```

Comment:
We apparently use `event.datetime`, that's creation time and not `received` I think: https://github.com/getsentry/sentry/blob/9c87c1ae59f6d1eae39e2bcecee0b192cb8bf3f5/src/sentry/event_manager.py#L1216

Reply:
fixed